# Data Analysis:

The data analysis process consists of the following steps:

1. <b>Asking Questions:</b> Varies to each scenaio of data we are dealing with.<br>
2. <b>Data Wrangling</b><br>
    a. <b>Gathering Data: </b>Collecting Data from different sources to work with.
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data: </b>To understand data and identify where the problem lies in data.<br>
    c. <b>Cleaning Data: </b>Removing problems found in data accessing.<br>
3. <b>Exploratory Data Analysis</b>
4. <b>Drawing Conclusion</b>
5. <b>Comunicating Results</b>
    

## <u>Assessing Data:</u>

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data (Data with Quality issues):</u> <br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data (Data with tidiness issues):</u><br>
    Messy data, also known as untidy data. Untidy data has structural issues.

#### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data.

#### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data.

### Types of Assessment

There are 2 types of assessment styles

1. <u>Manual:</u> Looking through the data manually in google sheets
2. <u>Programmatic:</u> By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues

# Performing Data Wrangling on Insulin DataSet

In [227]:
import pandas as pd
import re
import numpy as np

In [228]:
patients=pd.read_csv('Data/patients.csv')
treatment=pd.read_csv('Data/treatments.csv')
adverse=pd.read_csv('Data/adverse_reactions.csv')

In [229]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


## Step1: Perform Manual Analysis And Document The Results (For Quality Issue):

1. Focus on the most important columns:<br>
For example: <b>hba1c_change</b>, column in treament table.<br>
Problem in zipcode columns.

In [230]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [231]:
treatment

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [232]:
adverse

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

## Step2: Perform programmatic assessment (For Quality Issue).
1. Apply .info() to all tables without even thinking.<br>
<b>info()</b> function tells how many null values we have in dataframe and datatype for each column.
2. Apply describe() function to all tables.
3. Apply sample() function, randomly gives a row from dataframe.
4. Apply duplicated() function.

In [233]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [234]:
# Gives missing values from patients dataset cross checking with .info()
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [235]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [236]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [237]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [238]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [239]:
adverse.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top       cecilie  johnson     hypoglycemia
freq            1        2               19

In [240]:
patients.sample(5)

patient_id assigned_sex given_name      surname  \
368         369         male      Corey     Nicholls   
358         359       female     Amalie  Christensen   
85           86         male    Burrell      Mathieu   
127         128       female       Lena         Baer   
444         445         male   Angelico     Marchesi   

                        address           city       state  zip_code  \
368  3427 Gerald L. Bates Drive         Boston          MA    2110.0   
358          3077 Colony Street        Meriden          CT    6450.0   
85            3816 Grand Avenue      Kissimmee     Florida   34746.0   
127     272 Boone Crockett Lane           Elma          WA   98541.0   
444         456 Delaware Avenue  San Francisco  California   94108.0   

           country                                      contact   birthdate  \
368  United States   CoreyNicholls@jourrapide.com1 617 830 7216   4/25/1989   
358  United States  AmalieJChristensen@einrot.com1 203 235 1076   10/1/1941   
85   United States       407-662-9389BurrellMathieu@teleworm.us   5/16/1936   
127  United States               360-482-2553LenaBaer@rhyta.com  11/10/1977   
444  United States     415-277-2563AngelicoMarchesi@fleckens.hu  10/28/1945   

     weight  height   bmi  
368   165.0      74  21.2  
358   160.8      63  28.5  
85    203.5      71  28.4  
127   170.7      61  32.2  
444   211.6      69  31.2

In [241]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

## <u>Quality Issues:</u>

### Patients:
- zipcode column is stored in decimal format.
- Certain zipcodes are of 4 digits.
- For the state columns sometimes full names are used and other time short forms are used.
- missing values for contact details of 12 patients.
- Incorrect data types for assingned_sex and birth_date column.
- Errorneous data for weights and heights.
- Inconsistent way of storing phone numbers in contact rows.
- Duplicate data and not useful data in dataset.

### Treatments:
- hba1c_change col has missing values.
- Novodra and Auralin column has u attached to them.
- Only 280 rows are there instead of 350.
- Name and surname column is in lower case.
- Incorrect datatypefor auraling and novodra column.
- hba1c change 9 instead of 4.
- Use of dashes instead of nan in aurolin and novodra column.


### Adverse Reactions:
- Name and surname cols is in lower.

## <u>Tideness (Structural Problem):</u>

### Patients:
- ~~phone and email together in one col i.e contact column.~~

### Treatments:
- ~~3 variables stored in auraling and novodra columns.~~

### Adverse Reactions:
- ~~No point of using this table.~~

## Divide the qualitiy issues in following dimensions:

1. Completeness Issue
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues

### Patients:
- ~~zipcode column is stored in decimal format. [validity]~~
- ~~Certain zipcodes are of 4 digits. [validity]~~
- ~~For the state columns sometimes full names are used and other time short forms are used. [inconsistent data]~~
- missing values for contact details of 12 patients. [completeness issues]
- ~~Incorrect data types for assingned_sex and birth_date column. [?]~~
- Errorneous data for weights and heights. [accuracy issues]
- ~~Inconsistent way of storing phone numbers in contact rows. [inconsistent]~~
- ~~Duplicate data and not useful data in dataset. [completeness issue]~~

### Treatments:
- ~~hba1c_change col has missing values. [completeness issues]~~
- ~~Novodra and Auralin column has u attached to them. [validity issues]~~
- ~~Only 280 rows are there instead of 350. [completeness issues]~~
- ~~Name and surname column is in lower case. [consistency issues]~~
- ~~Incorrect datatypefor auraling and novodra column. [validity issues]~~
- ~~hba1c change 9 instead of 4. [accuracy issues]~~
- ~~Use of dashes instead of nan in aurolin and novodra column. [inconsistent issues]~~


### Adverse Reactions:
- ~~Name and surname cols is in lower. [inconsistent issues]~~

# Data Cleaning

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process.`

In [242]:
# Creating opies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()

## 1. Completeness Issues

- Missing values for contact details of 12 patients.
- Only 280 rows are there, instead of 350.
- hba1c_change col has missing values.
- Duplicate data and not useful data in dataset. (Patients)

#### Define

In [243]:
# Addining remaining data of treatment

#### Code

In [244]:
treatmentExtra = pd.read_csv("Data/treatments_cut.csv")
treatment_copy = treatment_copy.append(treatmentExtra)

#### Test

In [245]:
treatment_copy

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0    veronika    jindrová  41u - 48u          -         7.63       7.20   
1      elliot  richardson          -  40u - 45u         7.56       7.09   
2    yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3        skye  gormanston  33u - 36u          -         7.97       7.62   
4      alissa      montez          -  33u - 29u         7.78       7.46   
..        ...         ...        ...        ...          ...        ...   
65     rovzan     kishiev  32u - 37u          -         7.75       7.41   
66      jakob    jakobsen          -  28u - 26u         7.96       7.51   
67      bernd   schneider  48u - 56u          -         7.74       7.44   
68      berta  napolitani          -  42u - 44u         7.68       7.21   
69     armina       sauvé  36u - 46u          -         7.86       7.40   

    hba1c_change  
0            NaN  
1           0.97  
2            NaN  
3           0.35  
4           0.32  
..           ...  
65          0.34  
66          0.95  
67          0.30  
68           NaN  
69           NaN  

[350 rows x 7 columns]

#### Define

In [246]:
# Subtract hba1c_start from hba1c_end to get hba1c_change

#### Code

In [247]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

#### Test

In [248]:
treatment_copy.info()
treatment_copy.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


hba1c_start   hba1c_end  hba1c_change
count   350.000000  350.000000    350.000000
mean      7.956343    7.560057      0.396286
std       0.545328    0.545456      0.059283
min       7.500000    7.010000      0.200000
25%       7.650000    7.270000      0.350000
50%       7.785000    7.400000      0.390000
75%       7.950000    7.557500      0.440000
max       9.950000    9.580000      0.540000

#### Define

In [249]:
# Dealing with repeated data.

#### Code

In [250]:
patients_copy = patients_copy[~(patients_copy["contact"] == "johndoe@email.com1234567890")]

#### Test

In [251]:
patients_copy.shape

(497, 14)

## 2. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

#### Define

In [252]:
# Novodra and Aurolin using the melt function

#### Code

In [253]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [254]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str
treatment_copy.drop(columns={'dose'}, inplace=True)

/home/ritik/.local/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  


#### Test

In [255]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

#### Define

In [256]:
# Dealing with contact column

#### Code

In [257]:
email = []
phone = []
for i in patients_copy['contact']:
    if(str(i).lower() != 'nan'):
        pattern=re.compile(r'[a-zA-Z][a-zA-Z0-9.-]+[@][a-zA-Z0-9-]+\.[a-zA-Z0-9.-]+[a-zA-Z]')
        num = i.replace(pattern.findall(i)[0],'')
        num = str(num)[::-1]
        nums = re.findall(r'\d{3,4}',num)
        num = "+1 ("+str(nums[2])[::-1]+") "+str(nums[1])[::-1]+"-"+str(nums[0])[::-1]
        phone.append(num)
        
        email.append(pattern.findall(i)[0])
    else:
        phone.append(np.nan)
        email.append(np.nan)

In [258]:
patients_copy["phone"] = phone
patients_copy["email"] = email
patients_copy.drop(['contact'], axis = 1, inplace=True)

#### Test

In [259]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi              phone                      email  
0      66  19.6  +1 (951) 719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8  +1 (402) 363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1  +1 (334) 515-7487        TimNeudorf@cuvox.de

#### Define

In [260]:
# mergining adverse and treatment dataframe

#### Code

In [261]:
adverse_copy.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [262]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

In [263]:
treatment_copy = treatment_copy.merge(adverse_copy, how="left", on=["given_name", "surname"])

#### Test

In [264]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end adverse_reaction  
0  41u    48u              NaN  
1  33u    36u              NaN  
2  37u    42u              NaN  
3  31u    38u              NaN  
4  30u    36u              NaN

#### Define

In [265]:
# Dealing with Zip Code Problem

#### Code

In [266]:
zip = []
for i in patients_copy["zip_code"]:
    if(str(i).lower() != "nan"):
        code = str(int(i))
        if(len(code) != 5):
            code=code+"0"
        zip.append(code)
    else:
        zip.append(np.nan)

In [267]:
patients_copy["zip_code"] = zip

#### Test

In [268]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    70950  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi              phone                      email  
0      66  19.6  +1 (951) 719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8  +1 (402) 363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1  +1 (334) 515-7487        TimNeudorf@cuvox.de

#### Define

In [269]:
# Dealing with state column

#### Code

In [270]:
states = []
for i in patients_copy["state"]:
    if((str(i).lower() == "nan") | str(i).isupper()):
        states.append(np.nan)
    else:
        states.append(i)

In [271]:
patients_copy["state"] = states

#### Test

In [272]:
patients_copy

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state zip_code  \
0         576 Brown Bear Drive  Rancho California  California    92390   
1    2370 University Hill Road          Armstrong    Illinois    61812   
2        1493 Poling Farm Road               York    Nebraska    68467   
3          2335 Webster Street         Woodbridge         NaN    70950   
4         1428 Turkey Pen Lane             Dothan         NaN    36303   
..                         ...                ...         ...      ...   
498        2530 Victoria Court       Milton Mills         NaN    38520   
499   494 Clarksburg Park Road             Sedona         NaN    86341   
500          649 Nutter Street      Overland Park         NaN    64110   
501   3652 Boone Crockett Lane            Seattle         NaN    98109   
502          2778 North Avenue               Burr    Nebraska    68324   

           country  birthdate  weight  height   bmi              phone  \
0    United States  7/10/1976   121.7      66  19.6  +1 (951) 719-9170   
1    United States   4/3/1967   118.8      66  19.2  +1 (217) 569-3204   
2    United States  2/19/1980   177.8      71  24.8  +1 (402) 363-6804   
3    United States  7/26/1951   220.9      70  31.7  +1 (732) 636-8246   
4    United States  2/18/1928   192.3      27  26.1  +1 (334) 515-7487   
..             ...        ...     ...     ...   ...                ...   
498  United States  4/10/1959   181.1      72  24.6  +1 (207) 477-0579   
499  United States  3/26/1948   239.6      70  34.4  +1 (928) 284-4492   
500  United States  1/13/1971   171.2      67  26.8  +1 (816) 223-6007   
501  United States  2/13/1952   176.9      67  27.7  +1 (360) 443-2060   
502  United States   5/3/1954   138.2      71  19.3  +1 (402) 848-4923   

                                 email  
0             ZoeWellish@superrito.com  
1                 PamelaSHill@cuvox.de  
2                 JaeMDebord@gustr.com  
3            PhanBaLiem@jourrapide.com  
4                  TimNeudorf@cuvox.de  
..                                 ...  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com  

[497 rows x 15 columns]

#### Define

In [273]:
# Dealing with data type of assigned_sex and date and height

#### Code

In [274]:
patients_copy['assigned_sex'] = patients_copy.assigned_sex.astype('category')
patients_copy['height'] = patients_copy.height.astype('float')
patients_copy['birthdate'] = patients_copy.birthdate.astype('datetime64[ns]')

#### Test

In [275]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 497 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    497 non-null    int64         
 1   assigned_sex  497 non-null    category      
 2   given_name    497 non-null    object        
 3   surname       497 non-null    object        
 4   address       485 non-null    object        
 5   city          485 non-null    object        
 6   state         92 non-null     object        
 7   zip_code      485 non-null    object        
 8   country       485 non-null    object        
 9   birthdate     497 non-null    datetime64[ns]
 10  weight        497 non-null    float64       
 11  height        497 non-null    float64       
 12  bmi           497 non-null    float64       
 13  phone         485 non-null    object        
 14  email         485 non-null    object        
dtypes: category(1), datetime64[ns](1), float

#### Define

In [276]:
# Dealing with u attached to start and end

#### Code

In [277]:
start = []
end = []
for i in treatment_copy["start"]:
    start.append(i[0:-2])
for i in treatment_copy["end"]:
    end.append(i[0:-1].strip())

In [278]:
treatment_copy["start"] = start
treatment_copy["end"] = end

#### Test

In [279]:
treatment_copy

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  christopher    woodward         7.51       7.06          0.45   novodra   
346        maret    sultygov         7.67       7.30          0.37   novodra   
347        lixue       hsueh         9.21       8.80          0.41   novodra   
348        jakob    jakobsen         7.96       7.51          0.45   novodra   
349        berta  napolitani         7.68       7.21          0.47   novodra   

    start end           adverse_reaction  
0      41  48                        NaN  
1      33  36                        NaN  
2      37  42                        NaN  
3      31  38                        NaN  
4      30  36                        NaN  
..    ...  ..                        ...  
345    55  51                     nausea  
346    26  23                        NaN  
347    22  23  injection site discomfort  
348    28  26               hypoglycemia  
349    42  44  injection site discomfort  

[350 rows x 9 columns]

#### Define

In [280]:
# Making given_name and surname titlecase values

#### Code

In [282]:
treatment_copy['given_name'] = treatment_copy['given_name'].str.capitalize()
treatment_copy['surname'] = treatment_copy['surname'].str.capitalize()

#### Test

In [283]:
treatment_copy

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       Veronika    Jindrová         7.63       7.20          0.43   auralin   
1           Skye  Gormanston         7.97       7.62          0.35   auralin   
2         Sophia      Haugen         7.65       7.27          0.38   auralin   
3          Eddie      Archer         7.89       7.55          0.34   auralin   
4           Asia     Woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  Christopher    Woodward         7.51       7.06          0.45   novodra   
346        Maret    Sultygov         7.67       7.30          0.37   novodra   
347        Lixue       Hsueh         9.21       8.80          0.41   novodra   
348        Jakob    Jakobsen         7.96       7.51          0.45   novodra   
349        Berta  Napolitani         7.68       7.21          0.47   novodra   

    start end           adverse_reaction  
0      41  48                        NaN  
1      33  36                        NaN  
2      37  42                        NaN  
3      31  38                        NaN  
4      30  36                        NaN  
..    ...  ..                        ...  
345    55  51                     nausea  
346    26  23                        NaN  
347    22  23  injection site discomfort  
348    28  26               hypoglycemia  
349    42  44  injection site discomfort  

[350 rows x 9 columns]

#### Define

In [290]:
# Dealing with datatypes

#### Code

In [293]:
treatment_copy['treatment'] = treatment_copy.treatment.astype('category')
treatment_copy['start'] = treatment_copy.start.astype('float')
treatment_copy['end'] = treatment_copy.end.astype('float')

#### Test

In [294]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   given_name        350 non-null    object  
 1   surname           350 non-null    object  
 2   hba1c_start       350 non-null    float64 
 3   hba1c_end         350 non-null    float64 
 4   hba1c_change      350 non-null    float64 
 5   treatment         350 non-null    category
 6   start             350 non-null    float64 
 7   end               350 non-null    float64 
 8   adverse_reaction  35 non-null     object  
dtypes: category(1), float64(5), object(3)
memory usage: 25.0+ KB


# Checking If Everything Is Covered Or Not

In [295]:
patients_copy

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state zip_code  \
0         576 Brown Bear Drive  Rancho California  California    92390   
1    2370 University Hill Road          Armstrong    Illinois    61812   
2        1493 Poling Farm Road               York    Nebraska    68467   
3          2335 Webster Street         Woodbridge         NaN    70950   
4         1428 Turkey Pen Lane             Dothan         NaN    36303   
..                         ...                ...         ...      ...   
498        2530 Victoria Court       Milton Mills         NaN    38520   
499   494 Clarksburg Park Road             Sedona         NaN    86341   
500          649 Nutter Street      Overland Park         NaN    64110   
501   3652 Boone Crockett Lane            Seattle         NaN    98109   
502          2778 North Avenue               Burr    Nebraska    68324   

           country  birthdate  weight  height   bmi              phone  \
0    United States 1976-07-10   121.7    66.0  19.6  +1 (951) 719-9170   
1    United States 1967-04-03   118.8    66.0  19.2  +1 (217) 569-3204   
2    United States 1980-02-19   177.8    71.0  24.8  +1 (402) 363-6804   
3    United States 1951-07-26   220.9    70.0  31.7  +1 (732) 636-8246   
4    United States 1928-02-18   192.3    27.0  26.1  +1 (334) 515-7487   
..             ...        ...     ...     ...   ...                ...   
498  United States 1959-04-10   181.1    72.0  24.6  +1 (207) 477-0579   
499  United States 1948-03-26   239.6    70.0  34.4  +1 (928) 284-4492   
500  United States 1971-01-13   171.2    67.0  26.8  +1 (816) 223-6007   
501  United States 1952-02-13   176.9    67.0  27.7  +1 (360) 443-2060   
502  United States 1954-05-03   138.2    71.0  19.3  +1 (402) 848-4923   

                                 email  
0             ZoeWellish@superrito.com  
1                 PamelaSHill@cuvox.de  
2                 JaeMDebord@gustr.com  
3            PhanBaLiem@jourrapide.com  
4                  TimNeudorf@cuvox.de  
..                                 ...  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com  

[497 rows x 15 columns]

In [297]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 497 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    497 non-null    int64         
 1   assigned_sex  497 non-null    category      
 2   given_name    497 non-null    object        
 3   surname       497 non-null    object        
 4   address       485 non-null    object        
 5   city          485 non-null    object        
 6   state         92 non-null     object        
 7   zip_code      485 non-null    object        
 8   country       485 non-null    object        
 9   birthdate     497 non-null    datetime64[ns]
 10  weight        497 non-null    float64       
 11  height        497 non-null    float64       
 12  bmi           497 non-null    float64       
 13  phone         485 non-null    object        
 14  email         485 non-null    object        
dtypes: category(1), datetime64[ns](1), float

In [298]:
treatment_copy

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       Veronika    Jindrová         7.63       7.20          0.43   auralin   
1           Skye  Gormanston         7.97       7.62          0.35   auralin   
2         Sophia      Haugen         7.65       7.27          0.38   auralin   
3          Eddie      Archer         7.89       7.55          0.34   auralin   
4           Asia     Woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  Christopher    Woodward         7.51       7.06          0.45   novodra   
346        Maret    Sultygov         7.67       7.30          0.37   novodra   
347        Lixue       Hsueh         9.21       8.80          0.41   novodra   
348        Jakob    Jakobsen         7.96       7.51          0.45   novodra   
349        Berta  Napolitani         7.68       7.21          0.47   novodra   

     start   end           adverse_reaction  
0     41.0  48.0                        NaN  
1     33.0  36.0                        NaN  
2     37.0  42.0                        NaN  
3     31.0  38.0                        NaN  
4     30.0  36.0                        NaN  
..     ...   ...                        ...  
345   55.0  51.0                     nausea  
346   26.0  23.0                        NaN  
347   22.0  23.0  injection site discomfort  
348   28.0  26.0               hypoglycemia  
349   42.0  44.0  injection site discomfort  

[350 rows x 9 columns]

In [299]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   given_name        350 non-null    object  
 1   surname           350 non-null    object  
 2   hba1c_start       350 non-null    float64 
 3   hba1c_end         350 non-null    float64 
 4   hba1c_change      350 non-null    float64 
 5   treatment         350 non-null    category
 6   start             350 non-null    float64 
 7   end               350 non-null    float64 
 8   adverse_reaction  35 non-null     object  
dtypes: category(1), float64(5), object(3)
memory usage: 25.0+ KB


# Saving Our DataFrame

In [302]:
patients_copy.to_csv("Processed Data/patients.csv")
treatment_copy.to_csv("Processed Data/treatment.csv")